src/daft-image/src/ops.rs

```diff
@@ -359,3 +546,339 @@ pub fn fixed_image_html_value(arr: &FixedShapeImageArray, idx: usize, truncate:
+                // Difference (high frequency)
+                coeffs[y][x + current_size / 2] = (left - right) / 2.0;
+            }
+        }
+
+        // Vertical transform
+        for x in 0..current_size {
+            for y in 0..current_size / 2 {
+                let top = coeffs[y * 2][x];
+                let bottom = coeffs[y * 2 + 1][x];
+
+                // Average (low frequency)
+                coeffs[y][x] = f64::midpoint(top, bottom);
+                // Difference (high frequency)
+                coeffs[y + current_size / 2][x] = (top - bottom) / 2.0;
+            }
+        }
+
+        current_size /= 2;
+    }
+
+    // Extract 8x8 top-left corner (low frequency components)
+    let mut hash_bits = Vec::new();
+
+    // Calculate median of the 8x8 block
+    let mut values = Vec::new();
+    for y in 0..8 {
+        for x in 0..8 {
+            values.push(coeffs[y][x]);
+        }
+    }
+    values.sort_by(|a, b| a.partial_cmp(b).unwrap());
+    let median = if !values.is_empty() {
+        values[values.len() / 2]
+    } else {
+        0.0
+    };
+
+    // Generate hash bits
+    for y in 0..8 {
+        for x in 0..8 {
+            hash_bits.push(if coeffs[y][x] > median { '1' } else { '0' });
+        }
+    }
+
+    let hash = hash_bits.into_iter().collect::<String>();
+
+    // Validate length (safety check)
+    if hash.len() != 64 {
+        return Err(DaftError::ValueError(format!(
+            "Wavelet hash should be 64 characters, got {}",
+            hash.len()
+        )));
+    }
+
+    Ok(hash)
+}
+
+fn compute_crop_resistant_hash(img: CowImage) -> DaftResult<String> {
+    // Convert to grayscale
+    let gray_img = img.into_mode(daft_schema::prelude::ImageMode::L);
+
+    // Resize to 256x256 for crop-resistant analysis
+    let resized = gray_img.resize(256, 256);
+
+    // Get pixel data
+    let pixel_data = resized.as_u8_slice();
+
+    // Create multiple overlapping regions for analysis
+    // We'll create 8x8 regions at different positions to make it crop-resistant
+    let mut hash_bits = Vec::new();
+
+    // Define region positions (overlapping regions)
+    let region_size = 64;
+    let positions = [
+        (0, 0),     // Top-left
+        (96, 0),    // Top-right
+        (0, 96),    // Bottom-left
+        (96, 96),   // Bottom-right
+        (48, 48),   // Center
+        (32, 32),   // Upper-left center
+        (112, 32),  // Upper-right center
+        (32, 112),  // Lower-left center
+        (112, 112), // Lower-right center
+        (64, 64),   // True center
+        (16, 16),   // Top-left corner
+        (176, 16),  // Top-right corner
+        (16, 176),  // Bottom-left corner
+        (176, 176), // Bottom-right corner
+        (80, 80),   // Center-left
+        (176, 80),  // Center-right
+    ];
+
+    // Process each region
+    for (start_y, start_x) in &positions {
+        // Extract region
+        let mut region = Vec::new();
```

Comment:
perf: could you move this outside of the loop and manually clear it on each iteration. That way we're not allocating a vec on a loop inside of a loop.